cowprotocol/cowswap
# feat(permit): load pre generated permit info

# Summary

Part of https://github.com/cowprotocol/token-lists/issues/293

Depends on https://github.com/cowprotocol/token-lists/pull/299

Use the pre-generated permit info data.

1. When checking whether the token is permittable, check first on the pre-generated data
- Only when that's not available, check the token is permittable locally
2. In the token selector, use a combination of pre-generated data and locally fetched permit info
- With that info, show the flag in the dropdown selector
<img width="551" alt="image" src="https://github.com/cowprotocol/cowswap/assets/43217/639111bd-33c4-4908-89d7-0f735a3741a2">

<img width="550" alt="image" src="https://github.com/cowprotocol/cowswap/assets/43217/9876db2d-96e2-428c-8498-1344c1f8647d">

# To Test

1. Open the token selector
* There should be tokens with the gas free flag
2. Pick a permittable token you don't have allowance for
3. Check the network tab and filter for `PermitInfo`
* There should be a request for `PermitInfo.<chain id>.json` for the current network you are on
* Switching to different tokens, to another tab, switching to a different window should NOT trigger a new request of this file
4. Place order
* Should pop up to sign permit
* Order appData should contain permit pre hook
5. Load another token that's permittable and is not in the list, such as `0x30d20208d987713f46dfd34ef128bb16c404d10f` (SD) on mainnet
6. Pick it as a sell token
* Permit info for it should be loaded and stored on `permittableTokens:v1` localStorage
7. Open the token selector and search for the token that was just added
* Should have the gas free flag.
6. Swap it
* Should work like the one from the list
